## Do States with higher overall participation rates from year over year have better SAT and/or ACT outcomes?

## Executive Summary


### Contents:
- [2017 Data Import & Cleaning](#Data-Import-and-Cleaning)
- [2018 Data Import and Cleaning](#2018-Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-data)
- [Descriptive and Inferential Statistics](#Descriptive-and-Inferential-Statistics)
- [Outside Research](#Outside-Research)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

**If you combine your problem statement, executive summary, data dictionary, and conclusions/recommendations, you have an amazing README.md file that quickly aligns your audience to the contents of your project.** Don't forget to cite your data sources!

*All libraries used should be added here*

In [94]:
#Imports:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## 2017 Data Import and Cleaning

#### 1. Read In SAT & ACT  Data

Read in the `sat_2017.csv` and `act_2017.csv` files and assign them to appropriately named pandas dataframes.

In [95]:
#Code:
sat_2017 = pd.read_csv("../data/sat_2017.csv")
act_2017 = pd.read_csv("../data/act_2017.csv")

#### 2. Display Data

Print the first 10 rows of each dataframe to your jupyter notebook

In [96]:
sat_2017.shape

(51, 5)

In [97]:
#Code:
sat_2017.head(10)

State Participation  Evidence-Based Reading and Writing  \
0               Alabama            5%                                 593   
1                Alaska           38%                                 547   
2               Arizona           30%                                 563   
3              Arkansas            3%                                 614   
4            California           53%                                 531   
5              Colorado           11%                                 606   
6           Connecticut          100%                                 530   
7              Delaware          100%                                 503   
8  District of Columbia          100%                                 482   
9               Florida           83%                                 520   

   Math  Total  
0   572   1165  
1   533   1080  
2   553   1116  
3   594   1208  
4   524   1055  
5   595   1201  
6   512   1041  
7   492    996  
8   468    950  
9   497   1017

In [98]:
sat_2017.tail(10)

State Participation  Evidence-Based Reading and Writing  Math  \
41   South Dakota            3%                                 612   603   
42      Tennessee            5%                                 623   604   
43          Texas           62%                                 513   507   
44           Utah            3%                                 624   614   
45        Vermont           60%                                 562   551   
46       Virginia           65%                                 561   541   
47     Washington           64%                                 541   534   
48  West Virginia           14%                                 558   528   
49      Wisconsin            3%                                 642   649   
50        Wyoming            3%                                 626   604   

    Total  
41   1216  
42   1228  
43   1020  
44   1238  
45   1114  
46   1102  
47   1075  
48   1086  
49   1291  
50   1230

In [99]:
sat_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   State                               51 non-null     object
 1   Participation                       51 non-null     object
 2   Evidence-Based Reading and Writing  51 non-null     int64 
 3   Math                                51 non-null     int64 
 4   Total                               51 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


In [100]:
act_2017.shape

(52, 7)

In [101]:
act_2017.head(10)

State Participation  English  Math  Reading  Science  \
0              National           60%     20.3  20.7     21.4     21.0   
1               Alabama          100%     18.9  18.4     19.7     19.4   
2                Alaska           65%     18.7  19.8     20.4     19.9   
3               Arizona           62%     18.6  19.8     20.1     19.8   
4              Arkansas          100%     18.9  19.0     19.7     19.5   
5            California           31%     22.5  22.7     23.1     22.2   
6              Colorado          100%     20.1  20.3     21.2     20.9   
7           Connecticut           31%     25.5  24.6     25.6     24.6   
8              Delaware           18%     24.1  23.4     24.8     23.6   
9  District of Columbia           32%     24.4  23.5     24.9     23.5   

  Composite  
0      21.0  
1      19.2  
2      19.8  
3      19.7  
4      19.4  
5      22.8  
6      20.8  
7      25.2  
8      24.1  
9      24.2

In [102]:
act_2017.tail(10)

State Participation  English  Math  Reading  Science Composite
42   South Dakota           80%     20.7  21.5     22.3     22.0      21.8
43      Tennessee          100%     19.5  19.2     20.1     19.9      19.8
44          Texas           45%     19.5  20.7     21.1     20.9      20.7
45           Utah          100%     19.5  19.9     20.8     20.6      20.3
46        Vermont           29%     23.3  23.1     24.4     23.2      23.6
47       Virginia           29%     23.5  23.3     24.6     23.5      23.8
48     Washington           29%     20.9  21.9     22.1     22.0      21.9
49  West Virginia           69%     20.0  19.4     21.2     20.5      20.4
50      Wisconsin          100%     19.7  20.4     20.6     20.9      20.5
51        Wyoming          100%     19.4  19.8     20.8     20.6     20.2x

In [103]:
act_2017.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   State          52 non-null     object 
 1   Participation  52 non-null     object 
 2   English        52 non-null     float64
 3   Math           52 non-null     float64
 4   Reading        52 non-null     float64
 5   Science        52 non-null     float64
 6   Composite      52 non-null     object 
dtypes: float64(4), object(3)
memory usage: 3.0+ KB


#### 3. Verbally Describe Data

Take your time looking through the data and thoroughly describe the data in the markdown cell below. 

Answer: 

In the sat_2017 file, we see that it is detailing the breakdown of participation percentages and the mean of testing scores across 3 different categories, by state. The sat_2017 dataset has 5 columns. The 'State' and 'Participation' columns contain string object values, Patricipation presumably because they were initially entered in with the '%' sign. Every column appears to have 51 rows.

In the act_2017 file, we see that it shows the participation percentages and mean testing scores across 5 different testing categories, by both state <b>and</b> at the national level. There are 7 columns (2 more than the sat_2017 dataset). As with the previous dataset, the 'State' column and 'Participation' columns contains string object values, Patricipation presumably because they are displayed with the '%' sign. Every column has 52 rows. We may initially say that one reason for the 1+ column more than the sat_2017 dataset is due to the fact that the act_2017 has <i>national averages</i>

#### 4a. Does the data look complete? 

In [104]:
sat_2017.isna().sum()
act_2017.isna().sum()

State            0
Participation    0
English          0
Math             0
Reading          0
Science          0
Composite        0
dtype: int64

Answer:

Compared with one another, the overall row counts between the two datasets do not match. The ACT dataset has 1 more entry than the SAT. One explanation for this is a row in the ACT dataset specifically for national averages. This would indiate that the <b>SAT dataset may not be complete.</b>

Taken only in terms of one dataset in itself, given that each column contains the same number of rows, and no column in either dataset has any NaN / missing values, <b>the data does look to be complete intra-dataset</b>

#### 4b. Are there any obvious issues with the observations?

**What is the minimum *possible* value for each test/subtest? What is the maximum *possible* value?**

Consider comparing any questionable values to the sources of your data:
- [SAT](https://blog.collegevine.com/here-are-the-average-sat-scores-by-state/)
- [ACT](https://blog.prepscholar.com/act-scores-by-state-averages-highs-and-lows)

### Answer:

Minimum possible value for each test/subtest, SAT: <b>200 / Subtest</b>

Maximum possible value for each test/subtest, SAT: <b>800 / Subtest</b>
    
Minimum possible value for each test/subtest, ACT: <b>1 / Subtest</b>

Maximum possible value for each test/subtest, ACT: <b>36 / Subtest</b>

There is also a possible outlier sat_2017 Math column.

### The min in the sat_2017 column is well below the lowest possible score for the SAT

In [105]:
# print(sat_2017["Evidence-Based Reading and Writing"].describe())
print(sat_2017["Math"].describe())

# print(act_2017["English"].describe())
# print(act_2017["Math"].describe())
# print(act_2017["Reading"].describe())
# print(act_2017["Science"].describe())

count     51.000000
mean     547.627451
std       84.909119
min       52.000000
25%      522.000000
50%      548.000000
75%      599.000000
max      651.000000
Name: Math, dtype: float64


#### 4c. Fix any errors you identified

**The data is available** so there's no need to guess or calculate anything. If you didn't find any errors, continue to the next step.

In [106]:
#code
sat_2017["Math"].loc[20] # 20 is the cell that is throwing off the col
sat_2017.loc[20]

State                                 Maryland
Participation                              69%
Evidence-Based Reading and Writing         536
Math                                        52
Total                                     1060
Name: 20, dtype: object

#### 5. What are your data types? 
Display the data types of each feature. 

In [107]:
#code
sat_2017.dtypes

State                                 object
Participation                         object
Evidence-Based Reading and Writing     int64
Math                                   int64
Total                                  int64
dtype: object

In [108]:
act_2017.dtypes

State             object
Participation     object
English          float64
Math             float64
Reading          float64
Science          float64
Composite         object
dtype: object

What did you learn?
- Do any of them seem odd?  
- Which ones are not as they should be?  

### Answer:

The 'Participation' columns in both the sat_2017 and act_2017 datasets should not be a string type, we want to perform any calculations on it.
These should most likely be cast into floats.

The 'Composite' column in the act_2017 column is also a Series-Object / string. The values in this column are numeric, and should also most likely be cast into floats or integers.

#### 6. Fix Incorrect Data Types
Based on what you discovered above, use appropriate methods to re-type incorrectly typed data.
- Define a function that will allow you to convert participation rates to an appropriate numeric type. Use `map` or `apply` to change these columns in each dataframe.

In [109]:
#code
def format_participation(participation):
    try:
        return float(participation)
    except ValueError:
        participation = participation[:-1]
        return float(participation)

In [110]:
sat_2017["Participation"] = sat_2017["Participation"].apply(format_participation)
act_2017["Participation"] = act_2017["Participation"].apply(format_participation)

- Fix any individual values preventing other columns from being the appropriate type.

In [111]:
# Some values in actact_2017["Composite"] have alphabet characters in them.
# act_2017.tail(10)

In [112]:
# Check which values in Composite contain a non-numeric character:
act_2017["Composite"].value_counts() # Output = last cell in this col has the non-numeric character

20.3     3
19.8     3
21.4     3
23.6     2
19.4     2
21.8     2
24.2     2
24.1     2
21.9     2
20.4     2
19.7     2
25.2     1
22.6     1
20.8     1
23.9     1
19.0     1
23.8     1
24.3     1
20.5     1
20.2x    1
25.5     1
18.6     1
24.0     1
21.0     1
20.7     1
18.7     1
19.5     1
17.8     1
25.4     1
19.2     1
22.8     1
22.0     1
21.5     1
23.7     1
19.1     1
21.7     1
22.3     1
20.0     1
Name: Composite, dtype: int64

In [113]:
#code
def format_composite(composite):
    try:
        return float(composite)
    except ValueError:
        composite = composite[:-1]
        return float(composite)

In [114]:
act_2017["Composite"] = act_2017["Composite"].apply(format_composite)

In [115]:
# Check 
# act_2017.dtypes
# act_2017.tail()

- Finish your data modifications by making sure the columns are now typed appropriately.

In [116]:
#code
sat_2017.head()
act_2017.head()
sat_2017.tail()
act_2017.tail()

State  Participation  English  Math  Reading  Science  Composite
47       Virginia           29.0     23.5  23.3     24.6     23.5       23.8
48     Washington           29.0     20.9  21.9     22.1     22.0       21.9
49  West Virginia           69.0     20.0  19.4     21.2     20.5       20.4
50      Wisconsin          100.0     19.7  20.4     20.6     20.9       20.5
51        Wyoming          100.0     19.4  19.8     20.8     20.6       20.2

- Display the data types again to confirm they are correct.

In [117]:
#Code:
sat_2017.dtypes
act_2017.dtypes

State             object
Participation    float64
English          float64
Math             float64
Reading          float64
Science          float64
Composite        float64
dtype: object

#### 7. Rename Columns
Change the names of the columns to more expressive names so that you can tell the difference the SAT columns and the ACT columns. Your solution should map all column names being changed at once (no repeated singular name-changes). **We will be combining these data with some of the data from 2018, and so you should name columns in an appropriate way**.

**Guidelines**:
- Column names should be all lowercase (you will thank yourself when you start pushing data to SQL later in the course)
- Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`.
- Column names should be unique and informative (the only feature that we actually share between dataframes is the state).

In [118]:
#code
sat_2017_mapping = {"State": "state", 
               "Participation": "sat_2017_participation", 
               "Evidence-Based Reading and Writing": "sat_2017_reading_writing", 
               "Math": "sat_2017_math",
               "Total": "sat_2017_total"}

act_2017_mapping = {"State": "state", 
               "Participation": "act_2017_participation", 
               "English": "act_2017_english", 
               "Math": "act_2017_math", 
               "Reading": "act_2017_reading",
               "Science": "act_2017_science",
               "Composite": "act_2017_composite"}

sat_2017.rename(columns=sat_2017_mapping, inplace=True)
act_2017.rename(columns=act_2017_mapping, inplace=True)

#### 8. Create a data dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**

|Feature|Type|Dataset|Description|
|---|---|---|---|
|state|object|SAT|The state in which the test has been taken|
|participation|float|SAT|The percentage of students haven taken the SAT (Ex: 60.0 = 60%)|
|reading_writing|int|SAT|The average score earned in evident-based reading & writing by state|
|math|int|SAT|The average score earned by students in math by state|
|total|int|SAT|Total earned score acros both reading_writing and math by state|


|Feature|Type|Dataset|Description|
|---|---|---|---|
|state|object|ACT|The state in which the test has been taken|
|participation|float|ACT|The percentage of students haven taken the ACT (Ex 75.0 = 75%)|
|english|float|ACT|Average score earned in English subtest by state|
|math|float|ACT|Average score earned in Math subtest by state|
|reading|float|ACT|Average score earned in Reading subtest by state|
|science|float|ACT|Average score earned in Science subtest by state
|composite|float|ACT|Average score across English, Math, Reading, and Science by state|

#### 9. Drop unnecessary rows

One of our dataframes contains an extra row. Identify and remove this from the dataframe.

In [119]:
# sat_2017
# sat_2017["state"].duplicated()  # No duplicates in SAT

In [120]:
# act_2017  # Has 1 extra row than sat_2017
# act_2017["state"].duplicated()  # No duplicates in ACT
# Extra row in row 0

In [121]:
act_2017.head(2)

state  act_2017_participation  act_2017_english  act_2017_math  \
0  National                    60.0              20.3           20.7   
1   Alabama                   100.0              18.9           18.4   

   act_2017_reading  act_2017_science  act_2017_composite  
0              21.4              21.0                21.0  
1              19.7              19.4                19.2

In [122]:
#code
act_2017.drop(0, inplace=True)
act_2017.head(2)

state  act_2017_participation  act_2017_english  act_2017_math  \
1  Alabama                   100.0              18.9           18.4   
2   Alaska                    65.0              18.7           19.8   

   act_2017_reading  act_2017_science  act_2017_composite  
1              19.7              19.4                19.2  
2              20.4              19.9                19.8

In [123]:
act_2017.reset_index(drop=True, inplace=True)

#### 10. Merge Dataframes

Join the 2017 ACT and SAT dataframes using the state in each dataframe as the key. Assign this to a new variable.

In [124]:
#Code:
combined_2017_df = sat_2017.merge(act_2017, 
                                  how="outer", 
                                  on="state")

combined_2017_df.head(1)

state  sat_2017_participation  sat_2017_reading_writing  sat_2017_math  \
0  Alabama                     5.0                       593            572   

   sat_2017_total  act_2017_participation  act_2017_english  act_2017_math  \
0            1165                   100.0              18.9           18.4   

   act_2017_reading  act_2017_science  act_2017_composite  
0              19.7              19.4                19.2

#### 11. Save your cleaned, merged dataframe

Use a relative path to save out your data as `combined_2017.csv`.

In [125]:
#code
combined_2017_df.to_csv("../data/combined_2017.csv")

## 2018 Data Import and Cleaning

Read in the `sat_2018.csv` and `act_2018.csv` files and assign them to appropriately named pandas dataframes. For the **2018 ACT Data**, only the `Composite` scores are available. Repeat the same processes to clean the 2018 data here as you were instructed in the previous sections above.

In [126]:
# Read in Data
sat_2018 = pd.read_csv("../data/sat_2018.csv")
act_2018 = pd.read_csv("../data/act_2018.csv")

In [127]:
# Display Data
# sat_2018
# sat_2018.head(10)
# sat_2018.tail(10)
sat_2018.info()
# sat_2018.describe()
# sat_2018.dtypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   State                               51 non-null     object
 1   Participation                       51 non-null     object
 2   Evidence-Based Reading and Writing  51 non-null     int64 
 3   Math                                51 non-null     int64 
 4   Total                               51 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


In [128]:
# act_2018
# act_2018.head(10)
# act_2018.tail(10)
act_2018.info()
# act_2018.describe()
# act_2018.dtypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   State          52 non-null     object 
 1   Participation  52 non-null     object 
 2   Composite      52 non-null     float64
dtypes: float64(1), object(2)
memory usage: 1.3+ KB


## Verbally Describe Data
SAT 2018: 
The sat_2018 dataset contains a string column of state names, a string column of 'percentages' of participation by state, and 3 integer columns: 1 Evidence-Based Reading and Writing, 1 Math, and column of the totals of the previous two integer columns. None of the min or max values in any integer column seems like outside the boundaries of possible scores. Each Column has 51 non-null values.

ACT 2018:
The act_2018 dataset contains a string column of state names, a string column of 'percentages' of participation by state, and 1 integer column: the Composite score, which as average of a student's score across the 4 subtest types. None of the min or max values seems outside of the limits for scores. This dataset has 52 rows, 1 more than the sat_2018 dataset.

## A- Does the data look complete?
Neither the sat_2018 or act_2018 dataset has any NaN/missing values. This indicates that the data is complete.

The act_2018 DF does have one additional column that the sat_2018 DF does not have. The act_2018 DF is also missing columns that the act_2017 DF contained, indicating that, year over year, we are missing data. The act_2018 DF only contains the Participation and Composite columns for that year.

In [129]:
# sat_2018.isna().sum()
# act_2018.isna().sum()

## B- Are there any obvious issues with the data?
The Participation columns of both dataframes will need to be converted to an appropriate datatype (type float). The column names in both DF's will also need to be changed.

In [130]:
# C- Fix errors
# Find the reason why act_2018 has more rows:
act_2018[act_2018.duplicated()]

State Participation  Composite
20  Maine            7%       24.0

In [131]:
act_2018.drop(20, inplace=True)

In [132]:
act_2018.reset_index(drop=True, inplace=True)

In [133]:
# What are the dtypes?
sat_2018.dtypes
# act_2018.dtypes

State                                 object
Participation                         object
Evidence-Based Reading and Writing     int64
Math                                   int64
Total                                  int64
dtype: object

In [134]:
# Fix the wrong dtypes
sat_2018["Participation"] = sat_2018["Participation"].apply(format_participation)
act_2018["Participation"] = act_2018["Participation"].apply(format_participation)

In [135]:
# Rename cols
sat_2018.rename(columns={"State": "state",
                         "Participation": "sat_2018_participation",
                         "Evidence-Based Reading and Writing": "sat_2018_reading_writing",
                         "Math": "sat_2018_math",
                         "Total": "sat_2018_total"}, 
                         inplace=True)
act_2018.rename(columns={"State": "state",
                         "Participation": "act_2018_participation",
                         "Composite": "act_2018_composite"},
                         inplace=True)

# Data Dictionaries for SAT & ACT 2018
|Feature|Type|Dataset|Description|
|---|---|---|---|
|state|object|SAT|The state in which the test has been taken|
|participation|float|SAT|The percentage of students haven taken the SAT (Ex: 60.0 = 60%)|
|reading_writing|int|SAT|The average score earned in evident-based reading & writing by state|
|math|int|SAT|The average score earned by students in math by state|
|total|int|SAT|Total earned score acros both reading_writing and math by state|

|Feature|Type|Dataset|Description|
|---|---|---|---|
|state|object|ACT|The state in which the test has been taken|
|participation|float|ACT|The percentage of students haven taken the ACT (Ex 75.0 = 75%)|
|composite|float|ACT|Average score across English, Math, Reading, and Science by state|

In [136]:
# Merge DFs
combined_2018_df = sat_2018.merge(act_2018, on="state")
combined_2018_df.head(1)

state  sat_2018_participation  sat_2018_reading_writing  sat_2018_math  \
0  Alabama                     6.0                       595            571   

   sat_2018_total  act_2018_participation  act_2018_composite  
0            1166                   100.0                19.1

In [137]:
# Save the merge DF
combined_2018_df.to_csv("../data/combined_2018.csv")

#### Combine your 2017 and 2018 data into a single dataframe
Joining on state names should work, assuming you formatted all your state names identically. Make sure none of your columns (other than state) have identical names. Do yourself a favor and decide if you're encoding participation rates as floats or integers and standardize this across your datasets.

Save the contents of this merged dataframe as `final.csv`.

**Use this combined dataframe for the remainder of the project**.

In [138]:
final.to_csv("../data/final.csv")

NameError: name 'final' is not defined

In [ ]:
final = combined_2017_df.merge(combined_2018_df, on="state")

## Exploratory Data Analysis


### Summary Statistics
Transpose the output of pandas `describe` method to create a quick overview of each numeric feature.

In [ ]:
#Code:
final.describe()
# final.dtypes

#### Manually calculate standard deviation

$$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

- Write a function to calculate standard deviation using the formula above

In [ ]:
#code
# std = sqrt of sum(xi - mean)^2 / n

def calculate_std(df, col):
    squared_diffs = 0
    for i in df[col]:
        squared_diffs += ((i - df[col].mean()) ** 2)  
    return (squared_diffs / (df[col].count()-1)) ** 0.5
        

calculate_std(final, "sat_2017_reading_writing")

- Use a **dictionary comprehension** to apply your standard deviation function to each numeric column in the dataframe.  **No loops**  
- Assign the output to variable `sd` as a dictionary where: 
    - Each column name is now a key 
    - That standard deviation of the column is the value 
     
*Example Output :* `{'ACT_Math': 120, 'ACT_Reading': 120, ...}`

In [ ]:
#Code:
sd = {col: calculate_std(final, col) for col in final.iloc[:, 1:]}
sd

Do your manually calculated standard deviations match up with the output from pandas `describe`? What about numpy's `std` method?

In [ ]:
np.std(final["sat_2017_math"], ddof=1)

### Answer:
Initally, my manually calculated standard deviation did not match up with the output from pandas `describe`. When I accounted for bias (`n-1`), my manual standard deviation function matches the output 

Without using the `ddof=1`, numpy's `std` method, the output of the `std` method returned the same output as my own `calculate_std` function.

#### Investigate trends in the data
Using sorting and/or masking (along with the `.head` method to not print our entire dataframe), consider the following questions:

- Which states have the highest and lowest participation rates for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Which states have the highest and lowest mean total/composite scores for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Do any states with 100% participation on a given test have a rate change year-to-year?
- Do any states show have >50% participation on *both* tests either year?

Based on what you've just observed, have you identified any states that you're especially interested in? **Make a note of these and state *why* you think they're interesting**.

**You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

## The code below uses a variety of approaches to filter and sort data to return the relevant info

***The States with highest participation rates:***

- 2017 SAT: Connecticut (100%), Delaware (100%), Michigan (100%)
- 2018 SAT: Conneticut (100%), Colorado (100%), Delaware (100%), Idaho (100%), Michigan (100%)
- 2017 ACT: Alabama (100%), Montana (100%), Wisconsion (100%), Utah (100%), Tennessee (100%)
- 2018 ACT: Alabama (100%), Arkansas (100%), Wisconsin (100%), Utah (100%), Tennessee (100%)

In [ ]:
# Highest Participation by State:

high_sat_17_participation = final["sat_2017_participation"] > 75.
high_sat_18_participation = final["sat_2018_participation"] > 75.
high_act_17_participation = final["act_2017_participation"] > 75.
high_act_18_participation = final["act_2018_participation"] > 75.
final[high_sat_18_participation].sort_values("sat_2018_participation", ascending=False).head(10)

***The States with lowest participation rates:***

- 2017 SAT: Mississippi (2%), North Dakota (2%), Iowa (2%)
- 2018 SAT: North Dakota (2%)
- 2017 ACT: Maine (8%)
- 2018 ACT: Maine (7%)

In [ ]:
# Lowest Participation by State:

final[(final["sat_2017_participation"] < 10)].sort_values("sat_2017_participation", ascending=True).head()
final[(final["sat_2018_participation"] < 10)].sort_values("sat_2018_participation", ascending=True).head()
final[(final["act_2017_participation"] < 30)].sort_values("act_2017_participation", ascending=True).head()
final[(final["act_2017_participation"] < 30)].sort_values("act_2017_participation", ascending=True).head()

final["sat_2017_total"].sum()

***Highest Total (SAT) and Composite (ACT) Score by State:***

- 2017 SAT: Minnesota (1298)
- 2018 SAT: Minnesota (1295)
- 2017 ACT: New Hampshire (25.5)
- 2018 ACT: Conneticut (25.6)

In [ ]:
# Highest Mean Total/Composite Scores by State:

high_2017_total = (final["sat_2017_total"] == final["sat_2017_total"].max())
high_2018_total = (final["sat_2018_total"] == final["sat_2018_total"].max())
high_2017_compo = (final["act_2017_composite"] == final["act_2017_composite"].max())
high_2018_compo = (final["act_2018_composite"] == final["act_2018_composite"].max())
final[high_2018_compo].filter(["state", "sat_2017_total"])


***Lowest Total (SAT) and Composite (ACT) by State:***

- 2017 SAT: Delaware (996)
- 2018 SAT: Delaware (998)
- 2017 ACT: Nevada (17.8)
- 2018 ACT: Nevada (17.7)

In [ ]:
# Lowest Mean Total/Composite Scores by State:

# Credit to Alex Golden for showing me the df.filter() method to return only columns I want to observe while filtering/sorting.
final[(final["sat_2017_total"] == final["sat_2017_total"].min())].filter(["state", "sat_2017_total"])
final[(final["sat_2018_total"] == final["sat_2018_total"].min())].filter(["state", "sat_2018_total"])

final[(final["act_2017_composite"] == final["act_2017_composite"].min())].filter(["state", "act_2017_composite"])
final[(final["act_2018_composite"] == final["act_2018_composite"].min())].filter(["state", "act_2018_composite"])



In [ ]:
# Change in participation, change in score

***States with 100% participation on a given test have a rate change year-to-year?***

- Conneticut, Delware, and Michigan all had 100% participation in 2017, and had 100% participation in 2018. There was no change in the participation rate.

- 17 states had 100% participation ACT in 2017, and the ACT test in 2018 had 17 states with 100% participation as well.
    - Colorado and Minnesota had 100% participation in 2017, but did not in 2018.
    - Nebraska and Ohio did not have 100% participation in 2017, but did move up to 100% participation in 2018.

In [ ]:
sat_2017_full_particip = (final[final["sat_2017_participation"] == 100]).filter(("state", "sat_2017_participation"))
# print(sat_2017_full_particip)
sat_2018_full_particip = (final[final["sat_2018_participation"] == 100]).filter(("state", "sat_2017_participation"))
# print(sat_2018_full_particip)

In [ ]:
act_2017_full_particip = (final[final["act_2017_participation"] == 100]).filter(("state", "act_2017_participation"))
# print(act_2017_full_particip)
act_2018_full_particip = (final[final["act_2018_participation"] == 100]).filter(("state", "act_2018_participation"))
# print("\n", act_2018_full_particip)

***Do any states show have >50% participation on both tests either year?***

- 21 states had > 50% participation on the 2017 SAT.
- 25 states had > 50% participation on the 2018 SAT.
- Colorado, Illinois, North Caroline, and South Carolina all pushed above 50% participation in 2018 over 2017.

- 31 states had > 50% participation on the 2017 ACT.
- 28 states had > 50% participation on the 2018 ACT.
- Alaska, Colorado, Illinois all did not pass > 50% participation on the 2018 ACT, after reaching > 50% participation on the 2017 ACT.

In [ ]:
greater_sat_17_part = (final[final["sat_2017_participation"] > 50]).filter(("state", 
                                                                            "sat_2017_participation"))
print(greater_sat_17_part.count())
greater_sat_18_part = (final[final["sat_2018_participation"] > 50]).filter(("state", 
                                                                            "sat_2018_participation"))
print(greater_sat_18_part.count())

In [ ]:
final.loc[final["sat_2017_participation"] > 50]["state"]
final.loc[final["sat_2018_participation"] > 50]["state"]

In [ ]:
(final[final["act_2017_participation"] > 50]).filter(("state", 
                                                      " act_2017_participation")).count()
(final[final["act_2018_participation"] > 50]).filter(("state", 
                                                      "act_2018_participation")).count()

In [ ]:
print((final[final["act_2017_participation"] > 50])["state"])
print((final[final["act_2018_participation"] > 50])["state"])

## Visualize the data

There's not a magic bullet recommendation for the right number of plots to understand a given dataset, but visualizing your data is *always* a good idea. Not only does it allow you to quickly convey your findings (even if you have a non-technical audience), it will often reveal trends in your data that escaped you when you were looking only at numbers.

Some recommendations on plotting:
- Plots have titles
- Plots have axis labels
- Plots have appropriate tick labels
- All text is legible in a plot
- Plots demonstrate meaningful and valid relationships
- Plots are interpreted to aid understanding

There is such a thing as too many plots, and there are a *lot* of bad plots. You might make some! (But hopefully not with the guided prompts below).

#### Use Seaborn's heatmap with pandas `.corr()` to visualize correlations between all numeric features

Heatmaps are generally not appropriate for presentations, and should often be excluded from reports as they can be visually overwhelming. **However**, they can be extremely useful in identify relationships of potential interest (as well as identifying potential collinearity before modeling).

*example*:
```python
sns.heatmap(df.corr())
```

Please take time to format your output, adding a title. Look through some of the additional arguments and options. (Axis labels aren't really necessary, as long as the title is informative).

In [ ]:
final.corr()

In [ ]:
# Establish size of figure
plt.figure(figsize=(12, 8))
sns.heatmap(final.corr(),
            cmap="YlGnBu",
            annot=True)
plt.title("General Relation of Data Across All Categories", fontsize=20, pad=20.0);

In [ ]:
plt.figure(figsize=(12, 8))
sns.heatmap(final.iloc[:, :5].corr(),
            cmap="YlGnBu",
            annot=True)
plt.title("National Correlation 2017", fontsize=20, pad=20.0);

In [ ]:
plt.figure(figsize=(12, 8))
sns.heatmap(final.iloc[:, 10:14].corr(),
            cmap="coolwarm",
            annot=True)
plt.title("National Correlation 2018", fontsize=20, pad=20.0);

***Inside of each test year, and according the relevant test, the correlation tends to be quite strong between participation and subtest scores - although the SAT 2017, ACT 2017, SAT 2018, and ACT 2018 participation has a negative correlation to respective testing scores.***

#### Define a custom function to subplot histograms

We have data for two tests for two years. We only have composite (and not subtest scores) for the 2018 ACT. We should write a function that will take the names of 2+ columns and subplot histograms. While you can use pandas plotting or Seaborn here, matplotlib gives you greater control over all aspects of your plots.

[Helpful Link for Plotting Multiple Figures](https://matplotlib.org/users/pyplot_tutorial.html#working-with-multiple-figures-and-axes)

Here's some starter code:

In [ ]:
def subplot_histograms(dataframe, list_of_columns, list_of_titles, list_of_xlabels):
    nrows = int(np.ceil(len(list_of_columns)/2)) # Makes sure you have enough rows
    fig, ax = plt.subplots(nrows=nrows, ncols=2) # You'll want to specify your figsize
    ax = ax.ravel() # Ravel turns a matrix into a vector, which is easier to iterate
    for i, column in enumerate(list_of_columns): # Gives us an index value to get into all our lists
        ax[i].hist(dataframe[column]) # feel free to add more settings
        # Set titles, labels, etc here for each subplot
        plt[i].title(dataframe[column])
        

#### Plot and interpret histograms 
For each of the following:
- Participation rates for SAT & ACT
- Math scores for SAT & ACT
- Reading/verbal scores for SAT & ACT

In [ ]:
# Code
subplot_histograms(combined_2017_df, ["sat_2017_participation", "act_2017_participation"], 
                          ["SAT Participation", "ACT Participation"],
                           "X-label")

***Distributions of participation for SAT/ACT over both 2017 and 2018:***

In [ ]:
ax1 = plt.subplot(2, 1, 1)
ax1.set_title("SAT 2017 Participation")
ax1.hist(final["sat_2017_participation"])
ax1.vlines(x=np.mean(final["sat_2017_participation"]),
          ymin = 0,
          ymax = 15,
          color = "red");

ax2 = plt.subplot(2, 1, 2, sharex=ax1)
ax2.set_title("SAT 2018 Participation")
ax2.hist(final["sat_2018_participation"])
ax2.vlines(x=np.mean(final["sat_2018_participation"]),
          ymin = 0,
          ymax = 15,
          color = "red");
plt.tight_layout();

In [ ]:
plt.figure(figsize=(12, 12))

ax1 = plt.subplot(4, 1, 1)
ax1.set_title("ACT 2017 Participation")
ax1.hist(final["act_2017_participation"])
ax1.vlines(x=np.mean(final["act_2017_participation"]),
          ymin = 0,
          ymax = 15,
          color = "red");

ax2 = plt.subplot(4, 1, 2, sharex=ax1)  # sharex, as the values have the same possible ranges 
ax2.set_title("ACT 2018 Participation")
ax2.hist(final["act_2018_participation"])
ax2.vlines(x=np.mean(final["act_2018_participation"]),
          ymin = 0,
          ymax = 15,
          color = "red");

ax3 = plt.subplot(4, 1, 3, sharex=ax1)
ax3.set_title("SAT 2017 Participation")
ax3.hist(final["sat_2017_participation"])
ax3.vlines(x=np.mean(final["sat_2017_participation"]),
          ymin = 0,
          ymax = 15,
          color = "red");

ax4 = plt.subplot(4, 1, 4, sharex=ax1)
ax4.set_title("SAT 2018 Participation")
ax4.hist(final["sat_2018_participation"])
ax4.vlines(x=np.mean(final["sat_2018_participation"]),
          ymin = 0,
          ymax = 15,
          color = "red");

plt.tight_layout();

***Distributions of math scores for SAT/ACT available years:***

In [ ]:
plt.figure(figsize=(12, 8))

ax1 = plt.subplot(3, 1, 1)
ax1.set_title("ACT 2017 Math Scores")
ax1.hist(final["act_2017_math"])
ax1.vlines(x=np.mean(final["act_2017_math"]),
          ymin = 0,
          ymax = 15,
          color = "red");

ax2 = plt.subplot(3, 1, 2)  # If sharex == True, The scale of x axis is too small for the SAT score setup
ax2.set_title("All SAT 2017 Math Scores")
ax2.hist(final["sat_2017_math"])
ax2.vlines(x=np.mean(final["sat_2017_math"]),
          ymin = 0,
          ymax = 20,
          color = "red");

***Distributions of reading/verbal scores for SAT/ACT available years:***

In [ ]:
plt.figure(figsize=(12, 8))

ax1= plt.subplot(2, 1, 1)
ax1.set_title("ACT 2017 Reading/Verbal Scores")
ax1.hist(final["act_2017_reading"])
ax1.vlines(x=np.mean(final["act_2017_reading"]),
          ymin = 0,
          ymax = 12,
          color = "red");

ax2 = plt.subplot(2, 1, 2)
ax2.set_title("SAT 2017 Reading/Verbal Scores")
ax2.hist(final["sat_2017_reading_writing"])
ax2.vlines(x=np.mean(final["sat_2017_reading_writing"]),
          ymin = 0,
          ymax= 10,
          color = "red");

### Interpretation of Histograms

- It is difficult is some cases to compare the scores between the SAT and ACT: because of the different grading system, there isn't easy to find good jumping off points for measuring the differences between the two.
- That being said, the distribution (dist) for the ACT 2017 Math scores are fairly spread out. The SAT 2017 Math scores may be less spread, however, there is a likely outlier data point throwing off the dist visuals.

- The ACT 2017 Reading/Verbal scores are fairly spread out. The SAT 2017 Reading/Verbal scores have two 'peaks', indicating it could be bimonal dist.



#### Plot and interpret scatter plots

For each of the following:
- SAT vs. ACT math scores for 2017
- SAT vs. ACT verbal/reading scores for 2017
- SAT vs. ACT total/composite scores for 2017
- Total scores for SAT 2017 vs. 2018
- Composite scores for ACT 2017 vs. 2018

Plot the two variables against each other using matplotlib or Seaborn

Your plots should show:
- Two clearly labeled axes
- A proper title
- Using colors and symbols that are clear and unmistakable

**Feel free to write a custom function, and subplot if you'd like.** Functions save both time and space.


In [ ]:
# SAT v ACT Math Scores for 2017:
plt.figure(figsize=(12, 8))
plt.scatter(final["sat_2017_math"],
            final["act_2017_math"],
            color="orange");
plt.title("Math Scores 2017: SAT v. ACT")
plt.ylabel("ACT Math Scores", fontsize=15)
plt.xlabel("SAT Math Scores", fontsize=15);

In [ ]:
# SAT v ACT verbal scores for 2017:
plt.figure(figsize=(12, 8))
plt.scatter(final["sat_2017_reading_writing"],
            final["act_2017_reading"],
            c="blue");
plt.title("Verbal/Reading Scores 2017: SAT v. ACT")
plt.ylabel("ACT Reading Scores", fontsize=15)
plt.xlabel("SAT Reading/Verbal Scores", fontsize=15);

In [ ]:
# SAT v. AV total/composite scores for 2017
plt.figure(figsize=(12, 8))
plt.scatter(final["sat_2017_total"],
            final["act_2017_composite"],
            c="red")
plt.title("SAT Total v. ACT Composite 2017")
plt.ylabel("ACT Composite Scores", fontsize=15)
plt.xlabel("SAT Total Scores", fontsize=15);

In [ ]:
# Total scores for SAT 2017 vs 2018
plt.figure(figsize=(12, 8))
plt.scatter(final["sat_2017_total"],
            final["sat_2018_total"],
            c="brown")
plt.title("SAT Total Score 2017 v. 2018")
plt.ylabel("SAT Total Score 2018", fontsize=15)
plt.xlabel("SAT Total Score 2017", fontsize=15);

In [ ]:
# Composite scores for ACT 2017 vs. 2018
plt.figure(figsize=(12, 8))
plt.scatter(final["act_2017_composite"],
            final["act_2018_composite"],
            c="orange")
plt.title("ACT Composite Scores: 2017 v. 2018")
plt.ylabel("ACT 2018 Composite Scores", fontsize=15)
plt.xlabel("ACT 2017 Composite Scores", fontsize=15)

#### Plot and interpret boxplots

For each numeric variable in the dataframe create a boxplot using Seaborn. Boxplots demonstrate central tendency and spread in variables. In a certain sense, these are somewhat redundant with histograms, but you may be better able to identify clear outliers or differences in IQR, etc.

Multiple values can be plotted to a single boxplot as long as they are of the same relative scale (meaning they have similar min/max values).

Each boxplot should:
- Only include variables of a similar scale
- Have clear labels for each variable
- Have appropriate titles and labels

In [ ]:
final.describe()  # Find appropriate ranges to be compared in a boxplot

In [139]:
# All box plots of numeric values in the final DF:

# Did some research via stackoverflow to find something to make this work
# source URL: https://stackoverflow.com/questions/51777217/how-to-plot-a-boxplot-for-each-column-in-a-dataframe
for col in final.iloc[:, 1:]:
    plt.figure()  
    final.boxplot([col], color="red")

NameError: name 'final' is not defined

In [ ]:
plt.figure(figsize=(16, 4))
sns.boxplot(final["sat_2017_participation"], color="blue")

In [ ]:
plt.figure(figsize=(16, 4))
sns.boxplot(final["sat_2018_participation"], color="orange")

### Box Plot Interpretations and Observations:

- The SAT 2017 participation (partic) has quartiles on the lower end of the overall spread. The ACT 2017 partic is the inverse of that.

- Very little amount of spread indicated in the SAT 2017 Math box plot. THe same for ACT 2017 Science.

#### Feel free to do additional plots below
*(do research and choose your own chart types & variables)*

Are there any additional trends or relationships you haven't explored? Was there something interesting you saw that you'd like to dive further into? It's likely that there are a few more plots you might want to generate to support your narrative and recommendations that you are building toward. **As always, make sure you're interpreting your plots as you go**.

#### (Optional): Using Tableau, create a choropleth map for each variable using a map of the US. 

Save this plot as an image file in an images directory, provide a relative path, and insert the image into notebook in markdown.

## Outside Research

Based upon your observations, choose **three** states that demonstrate interesting trends in their SAT and/or ACT participation rates. Spend some time doing outside research on state policies that might influence these rates, and summarize your findings below. **Feel free to go back and create new plots that highlight these states of interest**. If you bring in any outside tables or charts, make sure you are explicit about having borrowed them. If you quote any text, make sure that it renders as being quoted. (Make sure that you cite your sources -- check with you local instructor for citation preferences).

***3 States to research:***

- Connecticut (high participation, high totals/composites)
    - https://portal.ct.gov/-/media/SDE/Student-Assessment/SAT/sat_faqs_superintendents.pdf
    - Gov Malloy sought and won approval for a waiver from US Dept of Ed to use SATs as the high school accountability measure
    - Starting late-2015, both in-person and technology based prof dev for teachers, students, parents, and others
    - accommodations for students with disabilities
    - Conn requires all 11th graders to take the SATs
- Minnesota (drop in participation %)
    - Very little info, appears to not have any if at all official policies regarding the taking of either SAT or ACT
- Delaware (High 2017 participation, low score total 2017)
    - https://www.udel.edu/apply/undergraduate-admissions/requirements/
    - For incoming state uni students, SAT is optional in admission decision
    - Seems primarily uni focused

## Conclusions and Recommendations

Based on your exploration of the data, what are you key takeaways and recommendations? Choose one state with a lower participation rate and provide a suggestion for how the College Board might increase participation amongst graduating seniors in this state. Are there additional data you desire that would better inform your investigations?

***Conclusions***

- One of the best examples of the relationship between participation and performance is Connecticut. With 100% participation year over year 2017-2018, this state has the highest composite score for 2018 (it should be noted that Connecticut scores were highest in ACT.

***Suggestions***

- Connecticut, a state with 100% year over year participation and high performance metrics, has moved forward with policies replacing state-issued standardized tests for the SAT. More policies like these will lead to better performance for incoming students.


***Additional Info Desired***

- In under standing the impact of state-wide participation on test outcomes, it would be useful to see participation data for the PSAT (or PreACT for ACT). Looking at participation in test-prep could shed more light on the effects of participation to score outcomes.

- What %'s of graduating high-school students that have taken the SAT or ACT have been accepted into 4 year universities? What about ones that have not? Data on whether students are in a position where the SAT or ACT is necesary for admissions could effect overall participation.

## Bonus: Descriptive and Inferential Statistics

#### Summarizing Distributions

Above, we used pandas `describe` to provide quick summary statistics of our numeric columns. We also demonstrated many visual relationships.

As data scientists, having a complete understanding of data is imperative prior to modeling.

While we will continue to build our analytic tools, we know that measures of *central tendency*, *spread*, and *shape/skewness* provide a quick summary of distributions.

For each variable in your data, summarize the underlying distributions (in words & statistics)
 - Be thorough in your verbal description of these distributions.
 - Be sure to back up these summaries with statistics.

Answers:

#### We generally assuming that data we sample from a population will be normally distributed. Do we observe this trend?

Answer:

Does This Assumption Hold for:
    - Math
    - Reading
    - Rates
Explain your answers for each distribution and how you think this will affect estimates made from these data.

Answer:

#### Estimate Limits of Data

Suppose we only seek to understand the relationship between SAT and ACT participation rates in 2017. 

##### Does it make sense to conduct statistical inference given these data specifically? 

Why or why not?

*(think about granularity, aggregation, the relationships between populations size & rates...consider the actually populations these data describe in answering this question)*

Answer:

##### Is it appropriate to compare *these* specific SAT and ACT math scores? 

Why or why not?

Answer:

#### Statistical Evaluation of Distributions 

**If you feel it's appropriate**, using methods we discussed in class, run hypothesis tests to compare variables of interest in our dataset. 

In [ ]:
# Code: